# import necessory libraries

In [1]:
import matplotlib.pyplot as plt
import matplotlib as mpl
import pandas as pd
import numpy as np

# import the dataset

In [2]:
df_alexa = pd.read_csv("amazon_alexa_data.csv")
df_alexa.head()

Unnamed: 0  rating       date         variation  \
0           0       5  31-Jul-18  Charcoal Fabric    
1           1       5  31-Jul-18  Charcoal Fabric    
2           2       4  31-Jul-18    Walnut Finish    
3           3       5  31-Jul-18  Charcoal Fabric    
4           4       5  31-Jul-18  Charcoal Fabric    

                                    verified_reviews  feedback  
0                                      Love my Echo!         1  
1                                          Loved it!         1  
2  Sometimes while playing a game, you can answer...         1  
3  I have had a lot of fun with this thing. My 4 ...         1  
4                                              Music         1

In [3]:
df_alexa.drop(df_alexa.columns[df_alexa.columns.str.contains('Unnamed',case = False)],axis = 1, inplace = True)

df_alexa.head()

rating       date         variation  \
0       5  31-Jul-18  Charcoal Fabric    
1       5  31-Jul-18  Charcoal Fabric    
2       4  31-Jul-18    Walnut Finish    
3       5  31-Jul-18  Charcoal Fabric    
4       5  31-Jul-18  Charcoal Fabric    

                                    verified_reviews  feedback  
0                                      Love my Echo!         1  
1                                          Loved it!         1  
2  Sometimes while playing a game, you can answer...         1  
3  I have had a lot of fun with this thing. My 4 ...         1  
4                                              Music         1

In [4]:
df_alexa.dropna(how= 'any', axis = 0)
df_alexa.head()

rating       date         variation  \
0       5  31-Jul-18  Charcoal Fabric    
1       5  31-Jul-18  Charcoal Fabric    
2       4  31-Jul-18    Walnut Finish    
3       5  31-Jul-18  Charcoal Fabric    
4       5  31-Jul-18  Charcoal Fabric    

                                    verified_reviews  feedback  
0                                      Love my Echo!         1  
1                                          Loved it!         1  
2  Sometimes while playing a game, you can answer...         1  
3  I have had a lot of fun with this thing. My 4 ...         1  
4                                              Music         1

In [5]:
df_alexa.isnull().sum()

rating              0
date                0
variation           0
verified_reviews    0
feedback            0
dtype: int64

In [6]:
df_alexa.variation.unique()

array(['Charcoal Fabric ', 'Walnut Finish ', 'Heather Gray Fabric ',
       'Sandstone Fabric ', 'Oak Finish ', 'Black', 'White',
       'Black  Spot', 'White  Spot', 'Black  Show', 'White  Show',
       'Black  Plus', 'White  Plus', 'Configuration: Fire TV Stick',
       'Black  Dot', 'White  Dot'], dtype=object)

In [10]:
from sklearn.preprocessing import OneHotEncoder
x_cat = df_alexa[['variation']]
encoder = OneHotEncoder()
x_cat = encoder.fit_transform(x_cat).toarray()
x_cat.shape

(3150, 16)

In [39]:

df2  = [df_alexa['verified_reviews'], df_alexa['feedback'], x_cat]
df_alexa= pd.concat([df_alexa,x_cat], axis=1)
df_alexa.drop(['variation'],axis=1,inplace=True)
df_alexa.head()

rating       date                                   verified_reviews  \
0       5  31-Jul-18                                      Love my Echo!   
1       5  31-Jul-18                                          Loved it!   
2       4  31-Jul-18  Sometimes while playing a game, you can answer...   
3       5  31-Jul-18  I have had a lot of fun with this thing. My 4 ...   
4       5  31-Jul-18                                              Music   

   feedback    0    1    2    3    4    5  ...    6    7    8    9   10   11  \
0         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   
1         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   
2         1  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  1.0  0.0   
3         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   
4         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   

    12   13   14   15  
0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  

[5 rows x 52 columns]

# tokenization of reviews

## import nltk for tokenization  

### convert  to lower case 

### remove punctuation

In [40]:
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords 
import string

df= np.asarray(df_alexa['verified_reviews'])
df_token=[]
for i in df:
    i= i.lower()  #convert  to lower case
    i = i.translate(str.maketrans('','',string.punctuation))   #removed punctuation 
    df_token.append(nltk.word_tokenize(i)) 



[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\Shree\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


### stopwords removed

### lemmatizing words

In [48]:
from nltk.stem import WordNetLemmatizer

nltk.download('wordnet')
nltk.download('omw-1.4')

stop_words = list(stopwords.words('english'))
df_token_filter = []

lemmatizer = WordNetLemmatizer()

for i in df_token:
    i = [x for x in i if x not in stop_words]
    i = [lemmatizer.lemmatize(x) for x in i]
    i = ' '.join(i)
    df_token_filter.append(i)
print(df_token_filter)

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\Shree\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\Shree\AppData\Roaming\nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


['love echo', 'loved', 'sometimes playing game answer question correctly alexa say got wrong answer like able turn light away home', 'lot fun thing 4 yr old learns dinosaur control light play game like category nice sound playing music well', 'music', 'received echo gift needed another bluetooth something play music easily accessible found smart speaker ’ wait see else', 'without cellphone use many feature ipad see use great alarm u r almost deaf hear alarm bedroom living room reason enough keep herit fun ask random question hear response seem smartbon politics yet', 'think 5th one ive purchased im working getting one every room house really like feature offer specifily playing music echo controlling light throughout house', 'look great', 'love ’ listened song ’ heard since childhood get news weather information ’ great', 'sent 85 year old dad talk constantly', 'love learning knew thing eveyday still figuring everything work far easy use understand make laugh time', 'purchased mother k

In [49]:
df_alexa['verified_reviews'] = df_token_filter


In [50]:
df_alexa.head()

rating       date                                   verified_reviews  \
0       5  31-Jul-18                                          love echo   
1       5  31-Jul-18                                              loved   
2       4  31-Jul-18  sometimes playing game answer question correct...   
3       5  31-Jul-18  lot fun thing 4 yr old learns dinosaur control...   
4       5  31-Jul-18                                              music   

   feedback    0    1    2    3    4    5  ...    6    7    8    9   10   11  \
0         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   
1         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   
2         1  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  1.0  0.0   
3         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   
4         1  0.0  0.0  0.0  0.0  0.0  1.0  ...  0.0  0.0  0.0  0.0  0.0  0.0   

    12   13   14   15  
0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  

[5 rows x 52 columns]

## transform words to vectors

In [54]:
from sklearn.feature_extraction.text import CountVectorizer
countv = CountVectorizer()

reviews_countv = countv.fit_transform(df_alexa['verified_reviews'])

reviews_countv.shape

(3150, 3989)

In [59]:
df_alexa.drop(['verified_reviews'], inplace =True , axis =1)


rating       date  feedback    0    1    2    3    4    5    6  ...    6  \
0       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   
1       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   
2       4  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
3       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   
4       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   

     7    8    9   10   11   12   13   14   15  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 51 columns]

In [60]:
df_alexa.head()

rating       date  feedback    0    1    2    3    4    5    6  ...    6  \
0       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   
1       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   
2       4  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0   
3       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   
4       5  31-Jul-18         1  0.0  0.0  0.0  0.0  0.0  1.0  0.0  ...  0.0   

     7    8    9   10   11   12   13   14   15  
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
2  0.0  0.0  0.0  1.0  0.0  0.0  0.0  0.0  0.0  
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[5 rows x 51 columns]

In [88]:
df_alexa2 = pd.DataFrame(reviews_countv.toarray())
df_alexa2

(3150, 3989)


0     1     2     3     4     5     6     7     8     9     ...  3979  \
0        0     0     0     0     0     0     0     0     0     0  ...     0   
1        0     0     0     0     0     0     0     0     0     0  ...     0   
2        0     0     0     0     0     0     0     0     0     0  ...     0   
3        0     0     0     0     0     0     0     0     0     0  ...     0   
4        0     0     0     0     0     0     0     0     0     0  ...     0   
...    ...   ...   ...   ...   ...   ...   ...   ...   ...   ...  ...   ...   
3145     0     0     0     0     0     0     0     0     0     0  ...     0   
3146     0     0     0     0     0     0     0     0     0     0  ...     0   
3147     0     0     0     0     0     0     0     0     0     0  ...     0   
3148     0     0     0     0     0     0     0     0     0     0  ...     0   
3149     0     0     0     0     0     0     0     0     0     0  ...     0   

      3980  3981  3982  3983  3984  3985  3986  3987  3988  
0        0     0     0     0     0     0     0     0     0  
1        0     0     0     0     0     0     0     0     0  
2        0     0     0     0     0     0     0     0     0  
3        1     0     0     0     0     0     0     0     0  
4        0     0     0     0     0     0     0     0     0  
...    ...   ...   ...   ...   ...   ...   ...   ...   ...  
3145     0     0     0     0     0     0     0     0     0  
3146     0     0     0     0     0     0     0     0     0  
3147     0     0     0     0     0     0     0     0     0  
3148     0     0     0     0     0     0     0     0     0  
3149     0     0     0     0     0     0     0     0     0  

[3150 rows x 3989 columns]

### determine X and y

In [84]:
X = df_alexa2
y = df_alexa['feedback']


(3150,)

### split test, train dataset

In [95]:
from sklearn.model_selection import train_test_split as tts

X_train,X_test, y_train , y_test = tts(X , y , test_size = 0.2)
print(X_train.shape, y_train.shape)

(2520, 3989) (2520,)


# implement Multinomial Naïve Bayes Classification

In [149]:
from sklearn.naive_bayes import MultinomialNB
clf = MultinomialNB().fit(X_train, y_train)

predicted = clf.predict(X_test)
predicted

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

In [150]:
from sklearn.metrics import accuracy_score as ac
from sklearn.metrics import confusion_matrix as cm

ac_mnb = ac(y_test, predicted)

print("accuracy score: " ,ac_mnb)
print("confusion matrix:  ")

cm(y_test, predicted)

accuracy score:  0.9349206349206349
confusion matrix:  


array([[ 21,  28],
       [ 13, 568]], dtype=int64)

# implement logistic regression

In [151]:
from sklearn.linear_model import LogisticRegression as lr
regr = lr()
regr = regr.fit(X_train,y_train)


In [152]:
predicted_lr = regr.predict(X_test)
predicted_lr

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,

In [153]:
ac_lr = ac(y_test, predicted_lr)

print("accuracy score: ", ac_lr)
print("confussion matrix: ")

cm(y_test, predicted_lr)

accuracy score:  0.9333333333333333
confussion matrix: 


array([[ 17,  32],
       [ 10, 571]], dtype=int64)

# implement KNN classifiers

In [110]:
from sklearn.neighbors import KNeighborsClassifier 


#### check how many neighbor produce best accuracy

In [154]:
Ks = 12
mean_acc = np.zeros((Ks-1))
std_acc = np.zeros((Ks-1))

for n in range(1,Ks):
    
    #Train Model and Predict  
    neigh = KNeighborsClassifier(n_neighbors = n).fit(X_train,y_train)
    yhat=neigh.predict(X_test)
    mean_acc[n-1] = ac(y_test, yhat)

    
    std_acc[n-1]=np.std(yhat==y_test)/np.sqrt(yhat.shape[0])

mean_acc

array([0.93809524, 0.93333333, 0.92222222, 0.92539683, 0.92222222,
       0.91904762, 0.92222222, 0.8984127 , 0.92222222, 0.9       ,
       0.92222222])

In [155]:
# from above test, neighbors = 5 looks optimal,so let's use it
knn = KNeighborsClassifier(n_neighbors = 5).fit(X_train,y_train)
predicted_knn = knn.predict(X_test)

ac_knn = ac(y_test ,predicted_knn)

print("accuracy score: ",ac_knn)
print("confusion matrix:")

cm(y_test, predicted_knn)

accuracy score:  0.9222222222222223
confusion matrix:


array([[  2,  47],
       [  2, 579]], dtype=int64)

### compare all classification modal accuracy

In [159]:
print("multinomialNB classifier test accuracy: ",ac_mnb)
print("logistic regression test accuracy: ",ac_lr)
print("K neighbour classifier test accuracy: ",ac_knn)    

multinomialNB classifier test accuracy:  0.9349206349206349
logistic regression test accuracy:  0.9333333333333333
K neighbour classifier test accuracy:  0.9222222222222223


### compare all confusion matrix

In [158]:
print("\nmultinomialNB classifier confision matrix: ")
print(cm(y_test,predicted))
print("\nlogistic regression confision matrix: ")
print(cm(y_test,predicted_lr))
print("\nK neighbour classifier confision matrix: ")
print(cm(y_test,predicted_knn))


multinomialNB classifier confision matrix: 
[[ 21  28]
 [ 13 568]]

logistic regression confision matrix: 
[[ 17  32]
 [ 10 571]]

K neighbour classifier confision matrix: 
[[  2  47]
 [  2 579]]


# from above comparisons I have derieved that  'Multinomial Naïve Bayes Classification' with accuracy = 93.49 % is the best classification model for sentiment prediction of amazon alexa review